bin/setup.ml

```diff
@@ -0,0 +1,114 @@
+(** Setup script to customize a new fsocaml project.
+
+    Patches to `dune-project` and the various `dune` files are done by parsing
+    the S-expressions and replacing instances of `fsocaml` when it's in a
+    non-label position in the `Sexp.List`.
+
+    Source file patching (e.g. `bin/main.ml` is done by searching and replacing
+    references to the Fsocaml module itself. *)
+
+open Core
+
+let dune_files = ["dune-project"; "test/dune"; "lib/dune"; "bin/dune"]
+
+let indent n =
+  String.init n ~f:(fun _ -> ' ')
+
+let pp_atom a =
+  if String.contains a ' ' then
+    "\"" ^ a ^ "\""
+  else a
+
+(** Convert an S-expression into a pretty-printed string in the style
+    of a dune configuration file. *)
+let pp_sexp s =
+  let rec loop s' i f =
+    match s' with
+    | Sexp.Atom a -> pp_atom a
+    | Sexp.List l ->
+        let nl = if f then "" else "\n" in
+        let prefix = sprintf "%s%s(" nl (indent (i * 2)) in
+        let ls = List.foldi l ~init:"" ~f:(fun j acc s'' -> 
+          let sp = if j = 0 then "" else " " in
+          acc ^ sp ^ (loop s'' (i + 1) false)
+        ) in
+        prefix ^ ls ^ ")"
+  in (loop s 0 true) ^ "\n"
+
+(** Scan an S-expression and replace all instances of `from` with `into`,
+    except for labels (the first atom in a list). *)
+let sexp_patch s from into =
+  let rec loop s' label  =
+    match s', label with
+    | Sexp.Atom a, false when (String.equal a from) -> Sexp.Atom into
+    | Sexp.Atom _, _ -> s'
+    | Sexp.List l, _ -> Sexp.List (List.mapi l ~f:(fun i s'' ->
+        if i = 0 then loop s'' true
+        else loop s'' false
+      ))
+  in loop s true
+
+let sexp_list_patch sl from into =
+  List.map sl ~f:(fun s ->
+    sexp_patch s from into
+  )
+
+let patch_sfile filename from into =
+  let sl = Sexp.load_sexps filename in
+  let f = (sexp_list_patch sl from into
+    |> List.map ~f:pp_sexp
+    |> String.concat ~sep:"\n"
+  ) in
+  Out_channel.with_file filename ~f:(fun file ->
+    Out_channel.output_string file f
+  )
```

Comment:
```suggestion
let patch_sfile filename from into =
  let f =
    Sexp.load_sexps filename
    |> List.map ~f:(sexp_patch ~from ~into)
    |> List.map ~f:pp_sexp
  in
  Out_channel.write_lines filename f
```

Reply:
Oh very nicely cleaned up – thank you! This is much more idiomatic